lib/link_state.h

```diff
@@ -0,0 +1,634 @@
+#define LS_PREF_UNSET		0x00
+#define LS_PREF_IGP_FLAG	0x01
+#define LS_PREF_ROUTE_TAG	0x02
+#define LS_PREF_EXTENDED_TAG	0x04
+#define LS_PREF_METRIC		0x08
+#define LS_PREF_SR		0x10
+
+/* Link State Prefix */
+struct ls_prefix {
+	uint8_t flags;			/* Flag for parameters validity */
+	struct ls_node_id adv;		/* Adv. Router of this Link State */
+	struct prefix pref;		/* IPv4 or IPv6 prefix */
+	uint8_t igp_flag;		/* IGP Flags associated to the prefix */
+	uint32_t route_tag;		/* IGP Route Tag */
+	uint64_t extended_tag;		/* IGP Extended Route Tag */
+	uint32_t metric;		/* Route metric for this prefix */
+	struct {
+		uint32_t sid;		/* Segment Routing ID */
+		uint8_t sid_flag;	/* Segment Routing Flags */
+		uint8_t algo;		/* Algorithm for Segment Routing */
+	} sr;
+};
+
+/**
+ * Create a new Link State Node. Structure is dynamically allocated.
+ *
+ * @param name	      Name of the router. May be NULL.
+ * @param router_id   Router ID as IPv4 address. Mandatory if router_id6 is 0
+ * @param router6_id  Router ID as IPv6 address. Mandatory if router_id is 0
+ * @param type        Type of router (Standard = 0 by default)
+ *
+ * @return            New Link State Node
+ */
+extern struct ls_node *ls_node_new(const char *name, struct in_addr router_id,
+				   struct in6_addr router6_id,
+				   enum node_type type);
+
+/**
+ * Remove Link State Node. Data structure is freed.
+ *
+ * @param node	      Pointer to a valid Link State Node structure
+ */
+extern void ls_node_del(struct ls_node *node);
+
+/**
+ * Create a new Link State Attributes. Structure is dynamically allocated.
+ * At least one of parameters MUST be valid and not equal to 0.
+ *
+ * @param local		Local IPv4 address
+ * @param local6	Local Ipv6 address
+ * @param local_id	Local Identifier
+ *
+ * @return		New Link State Attributes
+ */
+extern struct ls_attributes *ls_attributes_new(struct in_addr local,
+					       struct in6_addr local6,
+					       uint32_t local_id);
+
+/**
+ * Remove Link State Attributes. Data structure is freed.
+ *
+ * @param attr		 Pointer to a valid Link State Attribute structure
+ */
+extern void ls_attributes_del(struct ls_attributes *attr);
+
+
+/**
+ * In addition a Graph model is defined as an overlay on top of link state
+ * database in order to ease Path Computation algorithm implementation.
+ * Denoted G(V, E), a graph is composed by a list of Vertices (V) which
+ * represents the network Node and a list of Edges (E) which represents node
+ * Link. An additional list of prefixes (P) is also added.
+ * A prefix (P) is also attached to the Vertex (V) which advertise it.
+ *
+ * Vertex (V) contains the list of outgoing Edges (E) that connect this Vertex
+ * with its direct neighbors and the list of incoming Edges (E) that connect
+ * the direct neighbors to this Vertex. Indeed, the Edge (E) is unidirectional,
+ * thus, it is necessary to add 2 Edges to model a bidirectional relation
+ * between 2 Vertices.
+ *
```

Comment:
this is more of a question rather than a review comment ... I assume any process using this DB must perform any backcheck during SPF? Would it make sense to do the backcheck on insert and set a flag on the two "matching" entries?

I didn't see that anyplace, so thought I'd ask if this might make sense or not.

Reply:
DB is mostly here for Constraints SPF which in general does not need such backcheck. But, it could be useful for simple SPF. I'll look how to insert this for next release.